src/mongocxx/test/gridfs/bucket.cpp

```diff
@@ -1059,4 +1056,57 @@ TEST_CASE("gridfs download large file", "[gridfs::bucket]") {
 
     REQUIRE(total_bytes_read == length);
 }
+
+TEST_CASE("gridfs automatic index creation", "[gridfs::uploader] [gridfs::downloader]") {
+    using bsoncxx::builder::basic::document;
+    using bsoncxx::builder::basic::sub_array;
+
+    instance::current();
+    client client{uri{}};
+    database db = client["gridfs_index_creation"];
+    auto bucket = db.gridfs_bucket();
+
+    const size_t file_size = 100;
+    std::array<std::uint8_t, file_size> to_write, to_read;
+    std::iota(begin(to_write), end(to_write), 0);
+
+    auto uploader = bucket.open_upload_stream("test_file");
+
+    uploader.write(to_write.data(), file_size);
+    auto result = uploader.close();
+
+    auto downloader = bucket.open_download_stream(result.id());
+    auto bytes_read = downloader.read(to_read.data(), file_size);
+    REQUIRE(bytes_read == file_size);
+
+    SECTION("default indexes created") {
+        auto bucket_collection_name = GENERATE(values({"fs.chunks", "fs.files"}));
+        auto indexes = db[bucket_collection_name].list_indexes();
+        REQUIRE(std::distance(indexes.begin(), indexes.end()) == 2);
+    }
+
+    SECTION("fs.chunks 'files_id' and 'n' compares by numeric value") {
+        auto create_index_cmd =
+            make_document(kvp("createIndexes", "fs.chunks"), kvp("indexes", [&](sub_array sub_arr) {
+                              sub_arr.append(make_document(
+                                  kvp("key", make_document(kvp("_id", 1.0), kvp("n", 1.0))),
+                                  kvp("name", "files_id_1_n_1"),
+                                  kvp("unique", true)));
+                          }));
+        REQUIRE_THROWS_WITH(db.run_command(create_index_cmd.view()),
+                            Catch::Contains("must have unique name"));
```

Comment:
I'm expecting both indexes (`fs.chunks` and `fs.files`) to _not_ get created since they already exist. That should confirm that we're correctly matching the floating-point values here with the ints created by default. 

These do throw slightly different errors though and I wanted to point that out in case I'm doing anything wrong.

Reply:
I think the difference in server errors has to do with how server is comparing existing indexes.

As you noted, the gridfs bucket should have created this index on the chunks collection:
`{ "v" : 2, "unique" : true, "key" : { "files_id" : 1, "n" : 1 }, "name" : "files_id_1_n_1" }`

Since this is creating an index with a different key ` { _id: 1.0, n: 1.0 }`, but the same name. I think the server interprets that as an attempt to create a different index, since `_id` does not match `file_id`.

The test below attempts to create `{ filename: 1.0, uploadDate: 1.0 }`. The server may ignore the numeric type difference, and consider that an attempt to create the same exact index, which triggers the different error.

That's my guess. Checking in the shell seems to indicates the same. Running this twice:
```
db.runCommand({createIndexes: "coll", indexes: [{key: {x: 1}, name: "abc"}]})
```
Results in `all indexes already exist`

But creating a different index with the same name:
```
db.runCommand({createIndexes: "coll", indexes: [{key: {y: 1}, name: "index"}]})
```
Gets back `An existing index has the same name as the requested index`